### DISCUSSION TITLE
How to properly quote `[` in cmd?

### DISCUSSION THREAD

**Discussion by vitoyucepi** [nat-n/poethepoet#124] [Q&A] [answered]
## Summary

I've found https://github.com/nat-n/poethepoet/issues/60 and 2f2e6a39, but I didn't find any documentation about how I can escape `[`.

For example, I'd like to create a file that doesn't exist. Is it possible to use `if [ ! -f test ]; then touch test; fi`?
I can do this with the `test` instead.

**Comment by nat-n** [comment, marked as answer]
hi @vitoyucepi, 

Could you describe in a bit more detail what you're trying to do and how?

I'm not sure what the issue with escaping is, but I see you're trying to use shell syntax so you need to make sure the task is a shell task. The following should work:

```toml
setup.shell = "if [ ! -f test ]; then touch test; fi"
```

Also #12 might be interesting for you as well, though there's no ETA as of yet.

**Reply by vitoyucepi** [reply]
Let me start from the beginning.

I'm trying to create a task to build my project into a single file as youtube-dl did.
I researched a bit and found that it's possible to do this with a [shiv](https://github.com/linkedin/shiv) tool.
Unfortunately, shiv has a bug linkedin/shiv#231. Until this bug is fixed, I can try to do a workaround.
So I want to create a directory `dist` if it doesn't exist.

Overall my script looks like
```toml
[tool.poe.tasks]
pyz = [
  { cmd = "if [ ! -d dist ]; then mkdir dist; fi" },
  { cmd = """
  shiv
    -c main
    -o dist/project
    --compressed
    --reproducible
    --root ~/.cache/shiv
    .
  """ }
]
```

I run it with the command `poetry run poe pyz`.
In the log, I see that `[` is removed from the command `Poe => if ! -d dist ]; then mkdir dist; fi`.

**Reply by nat-n** [reply]
Thanks for explaining.

So I think my previous suggestion should help you here. I see what you've written here is a sequence task, with two inline cmd tasks. Unfortunately cmd tasks can't use shell syntax (because they run without a shell), so you need to replace the `cmd` key in the first task in the sequence with `shell` to tell poe to run it with a shell that can interpret the conditional syntax. i.e.

```toml
[tool.poe.tasks]
pyz = [
  { shell = "if [ ! -d dist ]; then mkdir dist; fi" },
  { cmd = """
  shiv
    -c main
    -o dist/project
    --compressed
    --reproducible
    --root ~/.cache/shiv
    .
  """ }
]
```

One caveat here is that because poe doesn't ship with a shell (that would be complicated to achieve) shell tasks only work if an appropriate shell interpreter (like bash or git-bash) can be found on the system which might not always be the case on windows.
